src/server/game/Combat/ThreatManager.h

```diff
@@ -15,284 +14,154 @@
-
-        void removeReference();
-
-        //=================================================
-
-        HostileReference* next() { return static_cast<HostileReference*>(Reference<Unit, ThreatManager>::next()); }
-
-        //=================================================
-
-        // Tell our refTo (target) object that we have a link
-        void targetObjectBuildLink() override;
-
-        // Tell our refTo (taget) object, that the link is cut
-        void targetObjectDestroyLink() override;
-
-        // Tell our refFrom (source) object, that the link is cut (Target destroyed)
-        void sourceObjectDestroyLink() override;
-
+        Unit* GetOwner() const { return _owner; }
+        Unit* GetVictim() const { return _victim; }
+        float GetThreat() const { return std::max<float>(_baseAmount + (float)_tempModifier, 0.0f); }
+        bool IsOnline() const { return (_online >= ONLINE_STATE_ONLINE); }
+        bool IsAvailable() const { return (_online > ONLINE_STATE_OFFLINE); }
+        bool IsOffline() const { return (_online <= ONLINE_STATE_OFFLINE); }
+
+        void SetThreat(float amount) { _baseAmount = amount; }
+        void AddThreat(float amount) { _baseAmount = std::max<float>(_baseAmount + amount, 0.0f); }
+        void ScaleThreat(float factor) { _baseAmount *= factor; }
+        void ModifyThreatByPercent(int32 percent) { ScaleThreat(0.01f*float(100 + percent)); }
+        void UpdateOnlineState();
+
+        void ClearThreat(bool sendRemove = true); // dealloc's this
     private:
-        // Inform the source, that the status of that reference was changed
-        void fireStatusChanged(ThreatRefStatusChangeEvent& threatRefStatusChangeEvent);
-
-        Unit* GetSourceUnit();
+        ThreatReference(Unit* owner, Unit* victim) : _owner(owner), _victim(victim), _baseAmount(0.0f), _tempModifier(0), _taunted(TAUNT_STATE_NONE), _online(ONLINE_STATE_OFFLINE) { }
+        OnlineState selectOnlineState();
+        void updateTauntState(bool victimIsTaunting);
+        Unit* const _owner;
+        Unit* const _victim;
+        float _baseAmount;
+        int32 _tempModifier; // Temporary effects (auras with SPELL_AURA_MOD_TOTAL_THREAT) - set from victim's threatmanager in ThreatManager::UpdateMyTempModifiers
+        TauntState _taunted;
+        OnlineState _online;
 
-    private:
-        float iThreat;
-        float iTempThreatModifier;                          // used for SPELL_AURA_MOD_TOTAL_THREAT
+    public:
+        ThreatReference(ThreatReference const&) = delete;
+        ThreatReference& operator=(ThreatReference const&) = delete;
 
-        ObjectGuid iUnitGuid;
-        bool iOnline;
-        bool iAccessible;
+    friend class ThreatManager;
+    friend struct CompareThreatLessThan;
 };
 
-//==============================================================
-class ThreatManager;
-
-class TC_GAME_API ThreatContainer
+struct CompareThreatLessThan : std::binary_function<ThreatReference*, ThreatReference*, bool>
```

Comment:
std::unary/binary_function structs are deprecated since C++11 and will be removed in C++17